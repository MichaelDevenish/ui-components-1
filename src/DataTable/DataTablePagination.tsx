import React from 'react'
import classNames from 'classnames'

export interface DataTablePaginationState {
  /** Currently selected page - stored here to allow async changes */
  page: number
}

export interface DataTablePaginationProps {
  /** Data to paginate - only used to get total number of rows */
  data: any[]

  /** Currently selected page */
  page: number

  /** Page count */
  pages: number

  /** Row size per page */
  pageSize: number

  /** List of page size options */
  pageSizeOptions: number[]

  /** Show the ability to change page size? */
  showPageSizeOptions: boolean

  /** Can we go to the next page? */
  canNext: boolean

  /** Can we go to the previous page? */
  canPrevious: boolean

  /** Classnames */
  className?: string

  /** Back button text */
  previousText?: string

  /** Forward button text */
  nextText?: string

  /** Page changing callback */
  onPageChange: (page: number) => void

  /** Page size change callback */
  onPageSizeChange: (newPageSize: number, newPage: number) => void
}

export class DataTablePagination extends React.Component<DataTablePaginationProps, DataTablePaginationState> {
  public static defaultProps: Partial<DataTablePaginationProps> = {
    previousText: 'Previous',
    nextText: 'Next'
  }

  constructor (props: DataTablePaginationProps) {
    super(props)

    this.state = {
      page: props.page
    }
  }

  componentWillReceiveProps (nextProps: DataTablePaginationProps) {
    this.setState({
      page: nextProps.page
    })
  }

  getSafePage = (newPage: number) => {
    const {
      pages
    } = this.props

    return Math.min(Math.max(newPage, 0), pages - 1)
  }

  changePage = (newPage: number) => {
    const {
      page,
      onPageChange
    } = this.props

    newPage = this.getSafePage(newPage)
    this.setState({ page: newPage })

    if (page !== newPage) {
      onPageChange(newPage)
    }
  }

  incrementPage = () => {
    const {
      page,
      canNext
    } = this.props

    if (canNext) {
      return this.changePage(page + 1)
    }
  }

  decrementPage = () => {
    const {
      page,
      canPrevious
    } = this.props

    if (canPrevious) {
      return this.changePage(page - 1)
    }
  }

  shouldPaginatePage = (pageNo: number) => {
    const { pages } = this.props
    const { page } = this.state

    // Always paginate start and end
    if (pageNo === 0 || pageNo + 1 === pages) {
      return true
    }

    // Paginate next and previous
    if (page - 1 <= pageNo && pageNo <= page + 1) {
      return true
    }

    // Paginate near start
    if (pageNo < 5 && page < 4) {
      return true
    }

    // Paginate near end
    if (pageNo > pages - 6 && page > pages - 5) {
      return true
    }

    return false
  }

  buttonForPage = (pageNo: number, key: number) => {
    const { page } = this.state

    return (
      <button
        key={key}
        className={classNames({'current': page === pageNo}, 'page-button', '-btn')}
        onClick={() => this.changePage(pageNo)}
      >
        {pageNo + 1}
      </button>
    )
  }

  get pagination () {
    const { pages } = this.props
    const { page } = this.state

    let response = []
    let hasStartEllipses = false
    let hasEndEllipses = false
    let key = 0 // Using a common key prevents jumpiness when changing pages
    for (let i = 0; i < pages; i++) {
      if (this.shouldPaginatePage(i)) {
        response.push(
          this.buttonForPage(i, key)
        )
        key++
      } else if (!hasStartEllipses && i < page) {
        response.push(
          <span className='ellipsis' key={key}>...</span>
        )
        hasStartEllipses = true
        key++
      } else if (!hasEndEllipses && i > page) {
        response.push(
          <span className='ellipsis' key={key}>...</span>
        )
        hasEndEllipses = true
        key++
      }
    }

    return response
  }

  get pageDetails () {
    const {
      pageSize,
      data
    } = this.props
    const { page } = this.state

    const entryCount = data.length
    const nextPage = page + 1
    const maxPossibleCurrentPageIndex = nextPage * pageSize
    const maxActualCurrentPageIndex = Math.min(entryCount, maxPossibleCurrentPageIndex)
    const pageStartingIndex = page * pageSize || 1

    return (
      <div className='page-details'>
        Showing {pageStartingIndex} to {maxActualCurrentPageIndex} of {entryCount} entries
      </div>
    )
  }

  get pageSizeOptions () {
    const {
      showPageSizeOptions,
      pageSizeOptions,
      pageSize,
      onPageSizeChange
    } = this.props
    const { page } = this.state

    if (showPageSizeOptions) {
      const pageSizeSelect = (
        <select
          onChange={e => onPageSizeChange(Number(e.target.value), page)}
          value={pageSize}
        >
          {pageSizeOptions.map((option: number) => (
            <option key={option} value={option}>
              {option}
            </option>
          ))}
        </select>
      )

      return (
        <span className='select-wrap -pageSizeOptions'>
          <label>Show {pageSizeSelect} entries</label>
        </span>
      )
    }
  }

  render () {
    const {
      canPrevious,
      canNext,
      className,
      previousText,
      nextText
    } = this.props

    return (
      <div className={classNames(className, '-pagination')} >
        {this.pageDetails}
        {this.pageSizeOptions}
        <div className='page-navigation'>
          <button
            type='button'
            className='previous -btn'
            onClick={this.decrementPage}
            disabled={!canPrevious}
          >
            {previousText}
          </button>
          {this.pagination}
          <button
            type='button'
            className='next -btn'
            onClick={this.incrementPage}
            disabled={!canNext}
          >
            {nextText}
          </button>
        </div>
      </div>
    )
  }
}
