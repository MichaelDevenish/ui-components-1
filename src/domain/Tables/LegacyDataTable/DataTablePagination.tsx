import classNames from 'classnames'
import React from 'react'

import { FontAwesomeIcon } from '../../Icons/FontAwesomeIcon'

export interface IDataTablePaginationState {
  /** Currently selected page - stored here to allow async changes */
  page: number
}

export interface IDataTablePaginationProps {
  totalCount: number
  /** Currently selected page */
  page: number
  /** Page count */
  pages: number
  /** Row size per page */
  pageSize: number
  /** List of page size options */
  pageSizeOptions: number[]
  /** Show the ability to change page size? */
  showPageSizeOptions: boolean
  /** Can we go to the next page? */
  canNext: boolean
  /** Can we go to the previous page? */
  canPrevious: boolean
  /** Classnames */
  className?: string
  /** Custom component to render within this component */
  customComponent?: JSX.Element
  /** Page changing callback */
  onPageChange: (page: number) => void
  /** Page size change callback */
  onPageSizeChange: (newPageSize: number, newPage: number) => void
}

export class DataTablePagination extends React.Component<IDataTablePaginationProps, IDataTablePaginationState> {
  constructor (props: IDataTablePaginationProps) {
    super(props)

    this.state = {
      page: props.page
    }
  }

  public UNSAFE_componentWillReceiveProps (nextProps: IDataTablePaginationProps) {
    this.setState({
      page: nextProps.page
    })
  }

  public getSafePage = (newPage: number) => {
    const {
      pages
    } = this.props

    return Math.min(Math.max(newPage, 0), pages - 1)
  }

  public changePage = (newPage: number) => {
    const {
      page,
      onPageChange
    } = this.props

    newPage = this.getSafePage(newPage)
    this.setState({ page: newPage })

    if (page !== newPage) {
      onPageChange(newPage)
    }
  }

  public incrementPage = () => {
    const {
      page,
      canNext
    } = this.props

    if (canNext) {
      return this.changePage(page + 1)
    }
  }

  public decrementPage = () => {
    const {
      page,
      canPrevious
    } = this.props

    if (canPrevious) {
      return this.changePage(page - 1)
    }
  }

  public shouldPaginatePage = (pageNo: number) => {
    const { pages } = this.props
    const { page } = this.state

    // Always paginate start and end
    if (pageNo === 0 || pageNo + 1 === pages) {
      return true
    }

    // Paginate next and previous
    if (page - 1 <= pageNo && pageNo <= page + 1) {
      return true
    }

    // Paginate near start
    if (pageNo < 5 && page < 4) {
      return true
    }

    // Paginate near end
    if (pageNo > pages - 6 && page > pages - 5) {
      return true
    }

    return false
  }

  public buttonForPage = (pageNo: number, key: number) => {
    const { page } = this.state

    const clickHandler = () => this.changePage(pageNo)

    return (
      <button
        key={key}
        className={classNames({ current: page === pageNo }, 'page-button', '-btn')}
        type='button'
        onClick={clickHandler}
      >
        {pageNo + 1}
      </button>
    )
  }

  get pagination () {
    const { pages } = this.props
    const { page } = this.state

    const response = []
    let hasStartEllipses = false
    let hasEndEllipses = false
    let key = 0 // Using a common key prevents jumpiness when changing pages
    for (let i = 0; i < pages; i++) {
      if (this.shouldPaginatePage(i)) {
        response.push(
          this.buttonForPage(i, key)
        )
        key++
      } else if (!hasStartEllipses && i < page) {
        response.push(
          <span className='ellipsis' key={key}>...</span>
        )
        hasStartEllipses = true
        key++
      } else if (!hasEndEllipses && i > page) {
        response.push(
          <span className='ellipsis' key={key}>...</span>
        )
        hasEndEllipses = true
        key++
      }
    }

    return response
  }

  get pageDetails () {
    const {
      pageSize,
      totalCount
    } = this.props
    const { page } = this.state

    const entryCount = totalCount
    const nextPage = page + 1
    const maxPossibleCurrentPageIndex = nextPage * pageSize
    const maxActualCurrentPageIndex = Math.min(entryCount, maxPossibleCurrentPageIndex)
    const pageStartingIndex = page * pageSize || 1

    return (
      <div className='page-details'>
        Showing {pageStartingIndex} to {maxActualCurrentPageIndex} of {entryCount} entries
      </div>
    )
  }

  get pageSizeOptions () {
    const {
      showPageSizeOptions,
      pageSizeOptions,
      pageSize
    } = this.props

    if (showPageSizeOptions) {
      const pageSizeSelect = (
        <select
          onChange={this.changeHandler}
          value={pageSize}
        >
          {pageSizeOptions.map((option: number) => (
            <option key={option} value={option}>
              {option}
            </option>
          ))}
        </select>
      )

      return (
        <span className='page-size-options'>
          <label>Show {pageSizeSelect} entries</label>
        </span>
      )
    }
  }

  public render () {
    const {
      canPrevious,
      canNext,
      className,
      customComponent
    } = this.props

    return (
      <div className={classNames(className, '-pagination')} >
        {customComponent}
        {this.pageSizeOptions}
        {this.pageDetails}
        <div className='page-navigation'>
          <button
            type='button'
            className='previous -btn'
            onClick={this.decrementPage}
            disabled={!canPrevious}
          >
            <FontAwesomeIcon type='solid' icon='chevron-left' />
          </button>
          {this.pagination}
          <button
            type='button'
            className='next -btn'
            onClick={this.incrementPage}
            disabled={!canNext}
          >
            <FontAwesomeIcon type='solid' icon='chevron-right' />
          </button>
        </div>
      </div>
    )
  }

  private changeHandler = (e: React.ChangeEvent<HTMLSelectElement>) => {
    const {
      onPageSizeChange
    } = this.props

    const {
      page
    } = this.state

    return onPageSizeChange(Number(e.target.value), page)
  }
}
