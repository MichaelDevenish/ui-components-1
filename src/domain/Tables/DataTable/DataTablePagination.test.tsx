import React from 'react'
import { shallow } from 'enzyme'
import { DataTablePagination } from './DataTablePagination'

describe('<DataTablePagination />', () => {
  describe('Pagination with few pages', () => {
    const onPageChange = jest.fn()
    const onPageSizeChange = jest.fn()
    const data = Array(70)

    const wrapper = shallow(
      <DataTablePagination
        data={data}
        pages={3}
        page={2}
        pageSize={25}
        pageSizeOptions={[5, 10, 20, 100]}
        showPageSizeOptions={false}
        canNext={false}
        canPrevious={true}
        onPageChange={onPageChange}
        onPageSizeChange={onPageSizeChange}
      />
    )

    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should not use ellipses for navigation', () => {
      expect(wrapper.find('.ellipsis').exists()).toBeFalsy()
    })

    it('should call the appropriate callbacks', () => {
      wrapper.find('.previous').simulate('click')

      expect(onPageChange).toBeCalledWith(1)
    })
  })

  describe('Pagination with many pages', () => {
    const onPageChange = jest.fn()
    const onPageSizeChange = jest.fn()
    const data = Array(295)

    const wrapper = shallow(
      <DataTablePagination
        data={data}
        pages={30}
        page={3}
        pageSize={10}
        pageSizeOptions={[5, 10, 20, 100]}
        showPageSizeOptions={false}
        canNext={true}
        canPrevious={true}
        onPageChange={onPageChange}
        onPageSizeChange={onPageSizeChange}
      />
    )

    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should collapse using ellipses for navigation', () => {
      expect(wrapper.find('.ellipsis').exists()).toBeTruthy()
    })
  })
})
