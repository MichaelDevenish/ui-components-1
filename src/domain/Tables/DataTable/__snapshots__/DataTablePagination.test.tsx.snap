// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DataTablePagination /> Pagination with few pages should match the snapshot 1`] = `
<div
  className="-pagination"
>
  <div
    className="page-details"
  >
    Showing 
    50
     to 
    70
     of 
    70
     entries
  </div>
  <div
    className="page-navigation"
  >
    <button
      className="previous -btn"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Previous
    </button>
    <button
      className="page-button -btn"
      key="0"
      onClick={[Function]}
    >
      1
    </button>
    <button
      className="page-button -btn"
      key="1"
      onClick={[Function]}
    >
      2
    </button>
    <button
      className="current page-button -btn"
      key="2"
      onClick={[Function]}
    >
      3
    </button>
    <button
      className="next -btn"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      Next
    </button>
  </div>
</div>
`;

exports[`<DataTablePagination /> Pagination with many pages should match the snapshot 1`] = `
<div
  className="-pagination"
>
  <div
    className="page-details"
  >
    Showing 
    30
     to 
    40
     of 
    295
     entries
  </div>
  <div
    className="page-navigation"
  >
    <button
      className="previous -btn"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Previous
    </button>
    <button
      className="page-button -btn"
      key="0"
      onClick={[Function]}
    >
      1
    </button>
    <button
      className="page-button -btn"
      key="1"
      onClick={[Function]}
    >
      2
    </button>
    <button
      className="page-button -btn"
      key="2"
      onClick={[Function]}
    >
      3
    </button>
    <button
      className="current page-button -btn"
      key="3"
      onClick={[Function]}
    >
      4
    </button>
    <button
      className="page-button -btn"
      key="4"
      onClick={[Function]}
    >
      5
    </button>
    <span
      className="ellipsis"
      key="5"
    >
      ...
    </span>
    <button
      className="page-button -btn"
      key="6"
      onClick={[Function]}
    >
      30
    </button>
    <button
      className="next -btn"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Next
    </button>
  </div>
</div>
`;
